### ISSUE TITLE
read: connection refused

### ISSUE THREAD

**Issue by SylvainBigonneau** [exercism/cli#390]
I installed the cli on Ubuntu 17.04 using the [install script](https://raw.githubusercontent.com/exercism/cli-www/master/public/install), and configured it using my API key.

Nonetheless, whether I try to do `exercism fetch javascript` or `exercism upgrade`, no matter on which network I am, I get an error of this type:

```
kenny@localhost:~$ exercism fetch javascript
2017/05/11 09:21:48 Get http://x.exercism.io/v2/exercises/javascript?key=something: dial tcp: lookup x.exercism.io on 127.0.1.1:53: read udp 127.0.0.1:58608->127.0.1.1:53: read: connection refused
```

This "port forwarding" thing looks sketchy to me... any idea what's going on there?

**Comment by Tonkpils** [comment]
`127.0.0.1` is the loopback IP which means you're essentially attempting to connect to your own computer. Could you show us `exercism debug`

**Comment by SylvainBigonneau** [comment]
Here is the output from `exercism debug`:

```
kenny@myHostName:~$ exercism debug

**** Debug Information ****
Exercism CLI Version: 2.4.0
2017/05/12 09:45:43 unable to fetch latest release: Get https://api.github.com/repos/exercism/cli/releases/latest: dial tcp: lookup api.github.com on 127.0.1.1:53: read udp 127.0.0.1:60943->127.0.1.1:53: read: connection refused
OS/Architecture: linux/amd64
Build OS/Architecture linux/amd64
Home Dir: /home/kenny
Config File: /home/kenny/.exercism.json
API Key: 8cd1*************************4d3
Exercises Directory: /home/kenny/exercism
Testing API endpoints reachability
	* GitHub API: https://api.github.com/ [Get https://api.github.com/: dial tcp: lookup api.github.com on 127.0.1.1:53: read udp 127.0.0.1:33037->127.0.1.1:53: read: connection refused] 1.105341ms
	* API: http://exercism.io [Get http://exercism.io: dial tcp: lookup exercism.io on 127.0.1.1:53: read udp 127.0.0.1:50516->127.0.1.1:53: read: connection refused] 1.294498ms
	* XAPI: http://x.exercism.io [Get http://x.exercism.io: dial tcp: lookup x.exercism.io on 127.0.1.1:53: read udp 127.0.0.1:50650->127.0.1.1:53: read: connection refused] 1.516803ms

If you are having trouble and need to file a GitHub issue (https://github.com/exercism/exercism.io/issues) please include this information (except your API key. Keep that private).
```

I thought my /etc/hosts file might have a problem too, but I don't see what...

```
kenny@myHostName:~$ cat /etc/hosts
127.0.0.1	localhost
127.0.1.1	myHostName

# The following lines are desirable for IPv6 capable hosts
::1     ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
```

**Comment by Tonkpils** [comment]
@SylvainBigonneau are you able to access the web from this computer? e.g `curl https://api.github.com/ -vv`

**Comment by SylvainBigonneau** [comment]
Yes, I can reach that no problem (am typing on this issue from the same computer). Thanks for the help!

```
kenny@myHostName:~$ curl https://api.github.com/ -vv
*   Trying 192.30.253.116...
* TCP_NODELAY set
* Connected to api.github.com (192.30.253.116) port 443 (#0)
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: C=US; ST=California; L=San Francisco; O=GitHub, Inc.; CN=*.github.com
*  start date: Jan 18 00:00:00 2017 GMT
*  expire date: Apr 17 12:00:00 2020 GMT
*  subjectAltName: host "api.github.com" matched cert's "*.github.com"
*  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=DigiCert SHA2 High Assurance Server CA
*  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: api.github.com
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: GitHub.com
< Date: Fri, 12 May 2017 13:36:54 GMT
< Content-Type: application/json; charset=utf-8
< Content-Length: 2165
< Status: 200 OK
< X-RateLimit-Limit: 60
< X-RateLimit-Remaining: 59
< X-RateLimit-Reset: 1494599814
< Cache-Control: public, max-age=60, s-maxage=60
< Vary: Accept
< ETag: "7dc470913f1fe9bb6c7355b50a0737bc"
< X-GitHub-Media-Type: github.v3; format=json
< Access-Control-Expose-Headers: ETag, Link, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval
< Access-Control-Allow-Origin: *
< Content-Security-Policy: default-src 'none'
< Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
< X-Content-Type-Options: nosniff
< X-Frame-Options: deny
< X-XSS-Protection: 1; mode=block
< Vary: Accept-Encoding
< X-Served-By: 075bb2f6b7031ca3c0e69edb17939fae
< X-GitHub-Request-Id: 4A86:4A1D:702B417:87D0185:5915BA76
< 
{
  "current_user_url": "https://api.github.com/user",
  "current_user_authorizations_html_url": "https://github.com/settings/connections/applications{/client_id}",
  "authorizations_url": "https://api.github.com/authorizations",
  "code_search_url": "https://api.github.com/search/code?q={query}{&page,per_page,sort,order}",
  "commit_search_url": "https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}",
  "emails_url": "https://api.github.com/user/emails",
  "emojis_url": "https://api.github.com/emojis",
  "events_url": "https://api.github.com/events",
  "feeds_url": "https://api.github.com/feeds",
  "followers_url": "https://api.github.com/user/followers",
  "following_url": "https://api.github.com/user/following{/target}",
  "gists_url": "https://api.github.com/gists{/gist_id}",
  "hub_url": "https://api.github.com/hub",
  "issue_search_url": "https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}",
  "issues_url": "https://api.github.com/issues",
  "keys_url": "https://api.github.com/user/keys",
  "notifications_url": "https://api.github.com/notifications",
  "organization_repositories_url": "https://api.github.com/orgs/{org}/repos{?type,page,per_page,sort}",
  "organization_url": "https://api.github.com/orgs/{org}",
  "public_gists_url": "https://api.github.com/gists/public",
  "rate_limit_url": "https://api.github.com/rate_limit",
  "repository_url": "https://api.github.com/repos/{owner}/{repo}",
  "repository_search_url": "https://api.github.com/search/repositories?q={query}{&page,per_page,sort,order}",
  "current_user_repositories_url": "https://api.github.com/user/repos{?type,page,per_page,sort}",
  "starred_url": "https://api.github.com/user/starred{/owner}{/repo}",
  "starred_gists_url": "https://api.github.com/gists/starred",
  "team_url": "https://api.github.com/teams",
  "user_url": "https://api.github.com/users/{user}",
  "user_organizations_url": "https://api.github.com/user/orgs",
  "user_repositories_url": "https://api.github.com/users/{user}/repos{?type,page,per_page,sort}",
  "user_search_url": "https://api.github.com/search/users?q={query}{&page,per_page,sort,order}"
}
* Curl_http_done: called premature == 0
* Connection #0 to host api.github.com left intact
```

**Comment by lcowell** [comment]
What are the contents of your `/etc/resolv.conf` ?

Do either of these commands work for you?

```
 dig +notcp api.github.com @127.0.0.1
 dig +tcp api.github.com @127.0.0.1
```

**Comment by antoniom13** [comment]
@SylvainBigonneau are you running this under a VM ? It seems that the shell you are running that command from is trying to query `127.0.0.1` for the `x.exercism.io` record.  

This is common in containerized environments where the DNS request (if left by default for most cases) is forwarded to the host machine.  If it's running in some type of virtualized environment then I would probably start there.  Not sure if there is a configuration issue or process you might need to change in order for the DNS requests to be resolved but it seems like it might not (?) be exercism ?

**Comment by SylvainBigonneau** [comment]
@antoniom13 Nope, plain old Ubuntu 17.04 on a Zenbook laptop.

@lcowell Here is the output from `cat resolv.conf`:

```
# Generated by NetworkManager
nameserver 127.0.1.1
```

The two `dig` commands:
```
kenny@myHostName:~$ dig +notcp api.github.com @127.0.0.1

; <<>> DiG 9.10.3-P4-Ubuntu <<>> +notcp api.github.com @127.0.0.1
;; global options: +cmd
;; connection timed out; no servers could be reached
```

```
kenny@myHostName:~$ dig +tcp api.github.com @127.0.0.1
;; Connection to 127.0.0.1#53(127.0.0.1) for api.github.com failed: connection refused.
```

**Comment by Tonkpils** [comment]
From what I gather and based on your OS, this may be an issue with your NetworkManager configuration and not with the CLI unfortunately

**Comment by lcowell** [comment]
@SylvainBigonneau do you have a `/etc/dnsmasq.conf` file? Can you post the contents?

**Comment by SylvainBigonneau** [comment]
@Tonkpils How bad... Litterally every other tools (firefox, chrome, thunderbird, npm, git, curl, wget, netflix, spotify, skype... you name it) work with my network except for this exercism client. :(

@lcowell No, but I have a `/etc/dnsmasq.d/network-manager` file:

```
# Tell any system-wide dnsmasq instance to make sure to bind to interfaces
# instead of listening on 0.0.0.0
# WARNING: changes to this file will get lost if network-manager is removed.
bind-interfaces
```

**Comment by Adrianacmy** [comment]
I have the same problem as @SylvainBigonneau, and got the same ouputs for above mentioned files, anyone knows how to solve it ? If it is the network manager, how should I change it?

**Comment by kytrinyx** [comment]
@Adrianacmy Would you please open a new issue where you run the same commands and copy/paste your output so that we can start fresh with your situation?

**Comment by Adrianacmy** [comment]
@kytrinyx, I updated my Ubuntu to 17.10, exercism is working now

**Comment by kytrinyx** [comment]
Glad to hear it, and thanks for reporting back!

**Comment by pchampin** [comment]
I'm using Ubuntu 17.10 and I'm still having the same problem as @SylvainBigonneau , with the same content for ``/etc/resolv.conf`` and ``/etc/dnsmasq.d/network-manager``, also the same failures with ``dig``... On the other hand ``systemd-resolve api.github.com`` replies with the correct IP address.

As I understand, systemd is providing its own name resolution service, which ``dig`` and ``exercism`` are note able to use... This is really strange since no other program seems to have any problem with it.

**Comment by kytrinyx** [comment]
@NobbZ do you know how to debug this?

**Comment by pchampin** [comment]
If that can help, below is the result of ``systemd-resolve --status``
```
$ systemd-resolve --status
Global
         DNS Servers: 127.0.1.1
          DNSSEC NTA: 10.in-addr.arpa
                      16.172.in-addr.arpa
                      168.192.in-addr.arpa
                      17.172.in-addr.arpa
                      18.172.in-addr.arpa
                      19.172.in-addr.arpa
                      20.172.in-addr.arpa
                      21.172.in-addr.arpa
                      22.172.in-addr.arpa
                      23.172.in-addr.arpa
                      24.172.in-addr.arpa
                      25.172.in-addr.arpa
                      26.172.in-addr.arpa
                      27.172.in-addr.arpa
                      28.172.in-addr.arpa
                      29.172.in-addr.arpa
                      30.172.in-addr.arpa
                      31.172.in-addr.arpa
                      corp
                      d.f.ip6.arpa
                      home
                      internal
                      intranet
                      lan
                      local
                      private
                      test

Link 4 (docker0)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 3 (br-9d94f874f233)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no

Link 2 (wlo1)
      Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
       LLMNR setting: yes
MulticastDNS setting: no
      DNSSEC setting: no
    DNSSEC supported: no
         DNS Servers: 192.168.0.254
                      2a01:e00::1
                      2a01:e00::2
```
So the "real" DNS server seems to be 192.168.0.254, and indeed ``dig api.github.com @192.168.0.254`` works fine.

**Comment by pchampin** [comment]
After more digging (no pun intended), the following issues seem related:
* https://github.com/kubernetes/kubernetes/issues/41975
* https://github.com/git-lfs/git-lfs/issues/2479
In both cases, the problem seem to be related to a VPN. I do have a VPN configured, but the problem is there regardless of me beeing connected to the VPN.

**Comment by kytrinyx** [comment]
@pchampin It does seem somewhat similar to the kubernetes issue, but it's odd that the problem would manifest even when not connected to the VPN. It looks like in the case of git-lfs, the problem happened only when _not_ connected to the VPN.

**Comment by pchampin** [comment]
Fixed it!
It seems that my Ubuntu 17.10 was badly configured (possibly the upgrade from previous versions wasn't perfect), and ``/etc/resolv.conf`` was not managed as it should. The solution was to remove ``/etc/resolv.conf`` (a regular file) and replace it by a symlink to ``/run/systemd/resolve/stub-resolv.conf``.

**Comment by kytrinyx** [comment]
@pchampin that's a relief! ⚡️ Thanks for reporting back how you solved it. That's always helpful the next time this comes up.

**Comment by NobbZ** [comment]
> @NobbZ do you know how to debug this?

Unfortunately not. I only realized in a different ticket the port number which I considered suspicious and therefore I said exactly that I assume a DNS configuration problem, but thats all I was able to tell back then or today as well.

Fortunately though @pchampin has found a solution which seems to work and I think this should be put into a FAQ or something, as this can get really tedious to debug and even worse seems to be not only distribution but also version dependent **and** how you installed that version (via upgrade or directly).
